cmd/derpprobe/derpprobe.go

```diff
@@ -135,13 +138,71 @@ func getOverallStatus() (o overallStatus) {
 	return
 }
 
+func notifySlack(text string) error {
+	type SlackRequestBody struct {
+		Text string `json:"text"`
+	}
+
+	slackBody, err := json.Marshal(SlackRequestBody{Text: text})
+	if err != nil {
+		return err
+	}
+
+	webhookUrl := os.Getenv("SLACK_WEBHOOK")
+	if webhookUrl == "" {
+		return errors.New("No SLACK_WEBHOOK configured")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, webhookUrl, bytes.NewBuffer(slackBody))
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Content-Type", "application/json")
+
+	// if Slack is down, don't break the DERP monitoring. Timeout quickly.
+	client := &http.Client{Timeout: 3 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(resp.Body)
+	if buf.String() != "ok" {
+		return errors.New("Non-ok response returned from Slack")
+	}
+	return nil
+}
+
 func serve(w http.ResponseWriter, r *http.Request) {
 	st := getOverallStatus()
 	summary := "All good"
-	if len(st.bad) > 0 {
+	if (float64(len(st.bad)) / float64(len(st.bad)+len(st.good))) > 0.25 {
+		// This will generate an alert and page a human.
+		// It also ends up in Slack, as part of the alert pipeline.
 		w.WriteHeader(500)
 		summary = fmt.Sprintf("%d problems", len(st.bad))
+	} else if len(st.bad) > 0 {
+		summary = fmt.Sprintf("%d problems", len(st.bad))
+		err := notifySlack(strings.Join(st.bad, "\n"))
```

Comment:
so this'll notify slack only when the HTML is fetched?

how about having one dedicated notify-slack goroutine running in a loop and notifying slack whenever a condition persists for over N seconds?

Reply:
Switched to a goroutine.